src/analysis/inner_product_space/finite_dimensional.lean

```diff
@@ -0,0 +1,354 @@
+/-
+Copyright (c) 2023 Monica Omar. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Monica Omar
+-/
+import linear_algebra.invariant_submodule
+import analysis.inner_product_space.adjoint
+import analysis.inner_product_space.spectrum
+
+/-!
+# Finite-dimensional inner product spaces
+
+In this file, we prove some results in finite-dimensional inner product spaces.
+
+## Notation
+
+This file uses the local notation `P _` for `orthogonal_projection _`
+and `↥P _` for the extended orthogonal projection `V →L[ℂ] V`.
+
+We let `V` be an inner product space over `ℂ`.
+-/
+
+variables {V : Type*} [inner_product_space ℂ V]
+
+local notation `P` := orthogonal_projection
+local notation `↥P` := orthogonal_projection.extend
+
+-- the extended orthogonal projection is an invariant subspace
+lemma submodule.invariant_of_ortho_proj (U : submodule ℂ V) [finite_dimensional ℂ V] :
+  submodule.invariant U (↥P U) := λ x hx, set_like.coe_mem (P U x : U)
+
+/-- if `U` is `T` invariant, then `(P U).comp T.comp (P U) = T.comp (P U)` -/
+lemma submodule.invariant_imp_ortho_proj_comp_T_comp_ortho_proj_eq_T_comp_ortho_proj
+  [finite_dimensional ℂ V] (U : submodule ℂ V) (T : V →ₗ[ℂ] V)
+  (h : submodule.invariant U T) (x : V) : ↑(P U (T ↑(P U x))) = T ↑(P U x) :=
+begin
+  obtain ⟨w, hw, v, hv, hvw⟩ := submodule.exists_sum_mem_mem_orthogonal U x,
+  rw [hvw, map_add,
+      orthogonal_projection_mem_subspace_orthogonal_complement_eq_zero hv,
+      add_zero, orthogonal_projection_eq_self_iff.mpr hw],
+  exact orthogonal_projection_eq_self_iff.mpr (h hw),
+end
+
+/-- if `(P U).comp T.comp (P U) = T.comp (P U)`, then `U` is `T` invariant  -/
+lemma submodule.ortho_proj_comp_T_comp_ortho_proj_eq_T_comp_ortho_proj_imp_invariant
+  [finite_dimensional ℂ V] (U : submodule ℂ V) (T : V →ₗ[ℂ] V)
+  (h : ∀ x : V, ↑(P U (T ↑(P U x))) = T ↑(P U x)) : submodule.invariant U T :=
+begin
+  intros u h_1,
```

Comment:
The name `h_1` looks a bit weird, you should use either `h1` or `h₁` or `hu` (because it's about u)